activerecord/test/cases/strict_test.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require "cases/helper"
+require "models/developer"
+require "models/computer"
+
+class StrictTest < ActiveRecord::TestCase
+  fixtures :developers
+
+  def test_strict
+    Developer.all.each { |d| assert_not d.strict? }
+    Developer.strict.each { |d| assert d.strict? }
+  end
+
+  def test_raises_if_strict_and_lazy_loading
+    dev = Developer.strict.first
+    assert_predicate dev, :strict?
+
+    assert_raises ActiveRecord::StrictViolationError do
+      dev.audit_logs.to_a
+    end
+  end
+
+  def test_preload_audit_logs_are_strict_because_parent_is_strict
+    developer = Developer.first
+
+    3.times do
+      AuditLog.create(developer: developer, message: "I am message")
+    end
+
+    dev = Developer.includes(:audit_logs).strict.first
+
+    assert_predicate dev, :strict?
+    assert dev.audit_logs.all?(&:strict?), "Expected all audit logs to be strict"
+  end
+end
```

Comment:
This could use more tests for relation methods.  For example, doing `dev.audit_logs.first` in `test_raises_if_strict_and_lazy_loading ` instead of `dev.audit_logs.to_a` will not result in an exception, which I find surprising.  If we are considering the relation itself to be strict, then I would expect that at least relation methods that use the records if they are `loaded?` to raise if the relation hasn't been loaded.

Reply:
I think you want the following to handle those methods on the collection proxy

```diff
diff --git a/activerecord/lib/active_record/associations/collection_association.rb b/activerecord/lib/active_record/associations/collection_association.rb
index f1b470f97e..958eba114d 100644
--- a/activerecord/lib/active_record/associations/collection_association.rb
+++ b/activerecord/lib/active_record/associations/collection_association.rb
@@ -306,6 +306,7 @@ def null_scope?
 
       def find_from_target?
         loaded? ||
+          owner.strict? ||
           owner.new_record? ||
           target.any? { |record| record.new_record? || record.changed? }
       end
diff --git a/activerecord/test/cases/strict_test.rb b/activerecord/test/cases/strict_test.rb
index 806810484f..5264521992 100644
--- a/activerecord/test/cases/strict_test.rb
+++ b/activerecord/test/cases/strict_test.rb
@@ -21,6 +21,15 @@ def test_raises_if_strict_and_lazy_loading
     end
   end
 
+  def test_raises_on_unloaded_relation_methods_if_strict
+    dev = Developer.strict.first
+    assert_predicate dev, :strict?
+
+    assert_raises ActiveRecord::StrictViolationError do
+      dev.audit_logs.first
+    end
+  end
+
   def test_preload_audit_logs_are_strict_because_parent_is_strict
     developer = Developer.first
 
```